pkg/image/controller/controller.go

```diff
@@ -0,0 +1,188 @@
+		return false
+	}
+	if len(repo.Tags) == 0 {
+		return true
+	}
+	emptyTags := 0
+	for _, v := range repo.Tags {
+		if len(v) == 0 {
+			emptyTags++
+		}
+	}
+	return emptyTags > 0
+}
+
+// Next processes the given image repository, looking for repos that have DockerImageRepository
+// set but have not yet been marked as "ready". If transient errors occur, err is returned but
+// the image repository is not modified (so it will be tried again later). If a permanent
+// failure occurs the image is marked with an annotation. The tags of the original spec image
+// are left as is (those are updated through status).
+func (c *ImportController) Next(repo *api.ImageRepository) error {
+	if !needsImport(repo) {
+		return nil
+	}
+	name := repo.DockerImageRepository
+
+	ref, err := api.ParseDockerImageReference(name)
+	if err != nil {
+		err = fmt.Errorf("invalid docker image repository, cannot import data: %v", err)
+		util.HandleError(err)
+		return c.done(repo, err.Error())
+	}
+
+	conn, err := c.client.Connect(ref.Registry)
+	if err != nil {
+		return err
+	}
+	tags, err := conn.ImageTags(ref.Namespace, ref.Name)
+	switch {
+	case dockerregistry.IsRepositoryNotFound(err), dockerregistry.IsRegistryNotFound(err):
+		return c.done(repo, err.Error())
+	case err != nil:
+		return err
+	}
+
+	newTags := make(map[string]string, len(repo.Tags))
+	imageToTag := make(map[string][]string)
+	switch {
+	case len(repo.Tags) == 0:
+		// copy all tags
+		for tag := range tags {
+			// TODO: switch to image when pull by ID is automatic
+			newTags[tag] = tag
+		}
+		for tag, image := range tags {
+			imageToTag[image] = append(imageToTag[image], tag)
+		}
+	default:
+		for tag, v := range repo.Tags {
+			if len(v) != 0 {
+				newTags[tag] = v
+				continue
+			}
+			image, ok := tags[tag]
+			if !ok {
+				// tag not found, set empty
+				continue
+			}
+			imageToTag[image] = append(imageToTag[image], tag)
+			// TODO: switch to image when pull by ID is automatic
+			newTags[tag] = tag
+		}
+	}
+
+	// nothing to tag - no images in the upstream repo, or we're in sync
+	if len(imageToTag) == 0 {
+		return c.done(repo, "")
+	}
+
+	for id, tags := range imageToTag {
+		dockerImage, err := conn.ImageByID(ref.Namespace, ref.Name, id)
+		switch {
+		case dockerregistry.IsRepositoryNotFound(err), dockerregistry.IsRegistryNotFound(err):
+			return c.done(repo, err.Error())
+		case dockerregistry.IsImageNotFound(err):
+			for _, tag := range tags {
+				delete(newTags, tag)
+			}
+			continue
+		case err != nil:
+			return err
+		}
+		var image api.DockerImage
+		if err := kapi.Scheme.Convert(dockerImage, &image); err != nil {
+			err = fmt.Errorf("could not convert image: %#v", err)
+			util.HandleError(err)
+			return c.done(repo, err.Error())
+		}
+
+		// if there is a tag for the image by its id (tag=tag), we can pull by id
+		tag := tags[0]
+		if hasTag(tags, id) {
+			tag = id
```

Comment:
Yeah, the tag is lost. For my repo https://registry.hub.docker.com/u/ncdc/testing/tags/manage/, you'll see 3 tags:

- latest
- a2
- 183b5f16d2e7562e477e35905310e21c356102f5dda55401d6737e446f7a082d

The "by id" tag (`183b...`) is for `a2`, and `latest` is different. In the yaml below, the first `status.tags` entry should have a tag of `a2`

```yaml
  status:
    dockerImageRepository: ncdc/testing
    tags:
    - items:
      - created: 2015-03-17T12:05:46-04:00
        dockerImageReference: ncdc/testing:183b5f16d2e7562e477e35905310e21c356102f5dda55401d6737e446f7a082d
        image: 183b5f16d2e7562e477e35905310e21c356102f5dda55401d6737e446f7a082d
      tag: 183b5f16d2e7562e477e35905310e21c356102f5dda55401d6737e446f7a082d
    - items:
      - created: 2015-03-17T12:05:46-04:00
        dockerImageReference: ncdc/testing:latest
        image: 5bb850d6904ceb7978c3c86d6022c6c8d807de072700aac32105b949e94a8539
      tag: latest
```